File: codes/python/wk03_angmom.py
"""
Week 3 - Angular Momentum Operators & Ladder Operators
Physical Chemistry 2, SKKU (Prof. S. Joon Kwon)

Builds Lz, L+, L-, Lx, Ly, L^2 as (2l+1)x(2l+1) matrices in the |l,m> basis
(hbar = 1) and verifies:
  [Lx, Ly] = i Lz (cyclic),  [L^2, Lz] = 0,
  L^2 = l(l+1) * Identity,
  L+|l,m> = sqrt((l-m)(l+m+1)) |l,m+1>,
  b_max = l, b_min = -l (ladder termination),
and the generalized uncertainty  sigma_Lx * sigma_Ly >= |<Lz>|/2.
"""
import numpy as np

def angmom_ops(l):
    """Return (Lx, Ly, Lz, Lp, Lm, L2) in the basis m = l, l-1, ..., -l."""
    dim = int(round(2 * l)) + 1
    ms = np.array([l - k for k in range(dim)], dtype=float)
    Lz = np.diag(ms)
    Lp = np.zeros((dim, dim))
    for k in range(1, dim):            # raises m: |l, m> -> |l, m+1>
        m = ms[k]
        Lp[k - 1, k] = np.sqrt(l * (l + 1) - m * (m + 1))
    Lm = Lp.T.copy()
    Lx = 0.5 * (Lp + Lm)
    Ly = -0.5j * (Lp - Lm)
    L2 = Lx @ Lx + Ly @ Ly + Lz @ Lz
    return Lx, Ly, Lz, Lp, Lm, L2

if __name__ == "__main__":
    for l in [1, 2]:
        Lx, Ly, Lz, Lp, Lm, L2 = angmom_ops(l)
        dim = Lz.shape[0]
        I = np.eye(dim)
        e1 = np.max(np.abs(Lx @ Ly - Ly @ Lx - 1j * Lz))   # [Lx,Ly] = i Lz
        e2 = np.max(np.abs(L2 @ Lz - Lz @ L2))             # [L^2,Lz] = 0
        e3 = np.max(np.abs(L2 - l * (l + 1) * I))          # L^2 = l(l+1) I
        print(f"l={l}: |[Lx,Ly]-iLz| = {e1:.2e},  |[L2,Lz]| = {e2:.2e},  |L2 - l(l+1)I| = {e3:.2e}")

    # Ladder coefficients for l = 2: c+(l,m) = sqrt((l-m)(l+m+1))
    l = 2
    _, _, _, Lp, Lm, _ = angmom_ops(l)
    print("c+(2,m) for m = 1, 0, -1, -2:",
          [f"{Lp[k - 1, k]:.4f}" for k in range(1, 5)],
          "  (exact: sqrt(4), sqrt(6), sqrt(6), sqrt(4) reversed order in m)")

    # Ladder termination: L+|l,l> = 0, L-|l,-l> = 0
    top = np.zeros(5); top[0] = 1.0     # |2, +2>
    bot = np.zeros(5); bot[-1] = 1.0    # |2, -2>
    print(f"|L+|2,+2>| = {np.linalg.norm(Lp @ top):.1e},  |L-|2,-2>| = {np.linalg.norm(Lm @ bot):.1e}")

    # Generalized uncertainty in |l=1, m=1>: sigma_Lx sigma_Ly >= |<Lz>|/2
    Lx, Ly, Lz, _, _, _ = angmom_ops(1)
    v = np.array([1.0, 0.0, 0.0])       # |1, +1>
    sx = np.sqrt(np.real(v @ (Lx @ Lx) @ v) - np.real(v @ Lx @ v)**2)
    sy = np.sqrt(np.real(v @ (Ly @ Ly) @ v) - np.real(v @ Ly @ v)**2)
    print(f"|1,+1>: sigma_Lx*sigma_Ly = {sx * sy:.4f} >= |<Lz>|/2 = {abs(np.real(v @ Lz @ v)) / 2:.4f}")

    # Vector model: |L| = sqrt(l(l+1)), cone half-angle for each m (l = 2)
    L = np.sqrt(2 * 3.0)
    for m in [2, 1, 0, -1, -2]:
        theta = np.degrees(np.arccos(m / L))
        print(f"l=2, m={m:+d}: |L| = sqrt(6) = {L:.4f} hbar, cone angle = {theta:.2f} deg")

File: codes/cpp/wk03_angmom.cpp
// Week 3 - Angular Momentum & Ladder Operators (|l,m> basis, hbar = 1)
// Physical Chemistry 2, SKKU (Prof. S. Joon Kwon)
// [Lx,Ly] = i Lz, [L^2,Lz] = 0, L^2 = l(l+1) I, L+|l,m> = c+(l,m)|l,m+1>
// Build: g++ -O2 -std=c++17 wk03_angmom.cpp -o wk03_angmom
#include <cstdio>
#include <cmath>
#include <complex>
#include <vector>

using cd = std::complex<double>;
using Mat = std::vector<std::vector<cd>>;

Mat zeros(int d) { return Mat(d, std::vector<cd>(d, 0.0)); }

Mat mul(const Mat& A, const Mat& B) {
    int d = (int)A.size(); Mat C = zeros(d);
    for (int i = 0; i < d; ++i)
        for (int k = 0; k < d; ++k)
            for (int j = 0; j < d; ++j) C[i][j] += A[i][k] * B[k][j];
    return C;
}

Mat sub(const Mat& A, const Mat& B) {
    int d = (int)A.size(); Mat C = zeros(d);
    for (int i = 0; i < d; ++i) for (int j = 0; j < d; ++j) C[i][j] = A[i][j] - B[i][j];
    return C;
}

double maxAbs(const Mat& A) {
    double m = 0; for (auto& r : A) for (auto& v : r) m = std::max(m, std::abs(v));
    return m;
}

struct AngMom { Mat Lx, Ly, Lz, Lp, Lm, L2; int d; };

AngMom angmom(int l) {                        // basis m = l, l-1, ..., -l
    int d = 2 * l + 1;
    AngMom A{zeros(d), zeros(d), zeros(d), zeros(d), zeros(d), zeros(d), d};
    for (int k = 0; k < d; ++k) A.Lz[k][k] = (double)(l - k);
    for (int k = 1; k < d; ++k) {
        double m = (double)(l - k);
        A.Lp[k - 1][k] = std::sqrt(l * (l + 1.0) - m * (m + 1.0));   // raises m
        A.Lm[k][k - 1] = A.Lp[k - 1][k].real();                      // = (L+)^T
    }
    for (int i = 0; i < d; ++i)
        for (int j = 0; j < d; ++j) {
            A.Lx[i][j] = 0.5 * (A.Lp[i][j] + A.Lm[i][j]);
            A.Ly[i][j] = cd(0.0, -0.5) * (A.Lp[i][j] - A.Lm[i][j]);
        }
    Mat t = mul(A.Lx, A.Lx), u = mul(A.Ly, A.Ly), w = mul(A.Lz, A.Lz);
    for (int i = 0; i < d; ++i)
        for (int j = 0; j < d; ++j) A.L2[i][j] = t[i][j] + u[i][j] + w[i][j];
    return A;
}

int main() {
    for (int l : {1, 2}) {
        AngMom A = angmom(l);
        Mat c1 = sub(mul(A.Lx, A.Ly), mul(A.Ly, A.Lx));   // should equal i Lz
        for (int i = 0; i < A.d; ++i)
            for (int j = 0; j < A.d; ++j) c1[i][j] -= cd(0.0, 1.0) * A.Lz[i][j];
        Mat c2 = sub(mul(A.L2, A.Lz), mul(A.Lz, A.L2));
        Mat c3 = A.L2;
        for (int i = 0; i < A.d; ++i) c3[i][i] -= (double)(l * (l + 1));
        std::printf("l=%d: |[Lx,Ly]-iLz| = %.2e, |[L2,Lz]| = %.2e, |L2-l(l+1)I| = %.2e\n",
                    l, maxAbs(c1), maxAbs(c2), maxAbs(c3));
    }

    // Ladder coefficients & termination for l = 2
    AngMom A = angmom(2);
    std::printf("c+(2,m) matrix elements:");
    for (int k = 1; k < A.d; ++k) std::printf(" %.4f", A.Lp[k - 1][k].real());
    std::printf("\nL+|2,+2> column norm = %.1e (termination)\n", std::abs(A.Lp[0][0]));

    // Vector model: |L| = sqrt(l(l+1)), cone angles for l = 2
    double L = std::sqrt(6.0);
    for (int m = 2; m >= -2; --m)
        std::printf("l=2, m=%+d: cone angle = %.2f deg\n", m,
                    std::acos(m / L) * 180.0 / M_PI);
    return 0;
}
